istioctl/pkg/kubernetes/pod.go

```diff
@@ -0,0 +1,195 @@
+			Container: container,
+			Command:   command,
+			Stdin:     false,
+			Stdout:    true,
+			Stderr:    true,
+			TTY:       false,
+		}, scheme.ParameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(client.Config, "POST", req.URL())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var stdout, stderr bytes.Buffer
+	err = exec.Stream(remotecommand.StreamOptions{
+		Stdin:  nil,
+		Stdout: &stdout,
+		Stderr: &stderr,
+		Tty:    false,
+	})
+
+	return &stdout, &stderr, err
+}
+
+// GetPilotAgentContainer retrieves the pilot-agent container name for the specified pod
+func (client *Client) GetPilotAgentContainer(podName, podNamespace string) (string, error) {
+	req := client.Get().
+		Resource("pods").
+		Namespace(podNamespace).
+		Name(podName)
+
+	res := req.Do()
+	if res.Error() != nil {
+		return "", fmt.Errorf("unable to retrieve Pod: %v", res.Error())
+	}
+	pod := &v1.Pod{}
+	if err := res.Into(pod); err != nil {
+		return "", fmt.Errorf("unable to parse Pod: %v", res.Error())
+	}
+	for _, c := range pod.Spec.Containers {
+		switch c.Name {
+		case "egressgateway", "ingress", "ingressgateway":
+			return c.Name, nil
+		}
+	}
+	return proxyContainer, nil
+}
+
+// CallPilotDiscoveryDebug calls the pilot-discover debug command for all pilots passed in with the provided information
+func (client *Client) CallPilotDiscoveryDebug(pods []v1.Pod, proxyID, configType string) (string, error) {
+	cmd := []string{"/usr/local/bin/pilot-discovery", "debug", proxyID, configType}
+	var err error
+	var stdout, stderr *bytes.Buffer
+	for _, pod := range pods {
+		stdout, stderr, err = client.PodExec(pod.Name, pod.Namespace, discoveryContainer, cmd)
+		if stdout.String() != "" {
+			if proxyID != mesh {
+				return stdout.String(), nil
+			}
+			fmt.Println(stdout.String())
+		}
+	}
+	if stderr.String() != "" {
+		return "", fmt.Errorf("unable to call pilot-discover debug: %v", stderr.String())
+	}
+	return stdout.String(), err
+}
+
+// CallPilotAgentDebug calls the pilot-agent debug command on the specified Pilot with the provided information
+func (client *Client) CallPilotAgentDebug(podName, podNamespace, configType string) (string, error) {
+	cmd := []string{"/usr/local/bin/pilot-agent", "debug", configType}
+	container, err := client.GetPilotAgentContainer(podName, podNamespace)
+	if err != nil {
+		return "", fmt.Errorf("unable to retrieve proxy container name: %v", err)
+	}
+	stdout, stderr, err := client.PodExec(podName, podNamespace, container, cmd)
+	if err != nil {
+		return stdout.String(), err
+	} else if stderr.String() != "" {
+		return "", fmt.Errorf("unable to call pilot-agent debug: %v", stderr.String())
+	}
+	return stdout.String(), nil
+}
+
+// CallPilotDiscoveryStatus calls the pilot-discover status command for all pilots passed in with the provided information
+func (client *Client) CallPilotDiscoveryStatus(pods []v1.Pod) ([][]byte, error) {
+	cmd := []string{"/usr/local/bin/pilot-discovery", "status"}
```

Comment:
I also not understand why we need to different `pilot-discovery status` (or `debug`) , which under the hood, make a http request to some debug url. Can we just have one command, e.g `pilot-discovery curl [url]` for everything?

Reply:
We can, yes. Although it will result in slightly less specific error messages.

Reply:
This is a very good point. I'll make the changes and hopefully reduce the amount of code.